src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -34624,16 +34633,50 @@ class D
     void Main(object? o)
     {
         boom();
+        o.ToString();
+
         [DoesNotReturn]
-        void boom()
-        {
-        }
+        void boom() { } // 1
     }
 }
 ";
-            // Should local functions support `[DoesNotReturn]`? https://github.com/dotnet/roslyn/issues/45814
             var comp = CreateNullableCompilation(new[] { source, DoesNotReturnAttributeDefinition }, parseOptions: TestOptions.Regular9);
-            comp.VerifyDiagnostics();
+            comp.VerifyDiagnostics(
+                // (11,21): warning CS8763: A method marked [DoesNotReturn] should not return.
+                //         void boom() { }
+                Diagnostic(ErrorCode.WRN_ShouldNotReturn, "{ }").WithLocation(11, 21)
+                );
+        }
+
+        [Fact]
+        [WorkItem(45814, "https://github.com/dotnet/roslyn/issues/45814")]
+        public void DoesNotReturnIf_LocalFunction()
+        {
+            string source = @"
+using System.Diagnostics.CodeAnalysis;
+class D
+{
+    void Main(object o, object? o2)
+    {
+        assert(o is null);
+        o.ToString(); // 1
+
+        assert(o2 is not null);
+        o2.ToString();
+
+        // Note: no enforcement on method body for DoesNotReturnIf
```

Comment:
Was this limitation intentional? Or the functionality was punted out?

Reply:
I think we failed to find a design for that enforcement (if I recall correctly it's not possible for a reasonable cost).